src/test/java/gregtech/common/covers/CoverFluidRegulatorTest.java

```diff
@@ -0,0 +1,227 @@
+		int amountTransferred = cfr.doKeepExact(1000, source, dest, isWater, 1000);
+
+		assertEquals("Wrong fluid amount moved", 1000, amountTransferred);
+	}
+
+	@Test
+	public void doKeepExact_moves_only_as_much_fluid_as_exists_in_the_source() {
+
+		// Create a regulator for testing with, and set it to "Keep Exact" mode
+		CoverFluidRegulator cfr = new CoverFluidRegulator(null, EnumFacing.UP, 0, 1000);
+		cfr.transferMode = TransferMode.KEEP_EXACT;
+
+		IFluidHandler source =
+			new FluidHandlerProxy(new FluidTankList(false),
+			                      new FluidTankList(false,
+			                                        new FluidTank(new FluidStack(FluidRegistry.WATER, 1234), 64000)));
+
+		IFluidHandler dest =
+			new FluidHandlerProxy(new FluidTankList(false, new FluidTank(64000)),
+			                      new FluidTankList(false));
+
+		int amountTransferred = cfr.doKeepExact(10000, source, dest, isWater, 10000);
+
+		assertEquals("Wrong fluid amount moved", 1234, amountTransferred);
+	}
+
+	@Test
+	public void doKeepExact_moves_only_the_fluid_required_if_more_could_be_moved() {
+
+		CoverFluidRegulator cfr = new CoverFluidRegulator(null, EnumFacing.UP, 0, 1000);
+		cfr.transferMode = TransferMode.KEEP_EXACT;
+
+		IFluidHandler source =
+			new FluidHandlerProxy(
+				new FluidTankList(false),
+				new FluidTankList(false,
+				                  new FluidTank(new FluidStack(FluidRegistry.WATER, 64000), 64000)));
+
+		IFluidHandler dest =
+			new FluidHandlerProxy(
+				new FluidTankList(false,
+				                  new FluidTank(new FluidStack(FluidRegistry.WATER, 100), 64000)),
+				new FluidTankList(false));
+
+		int amountTransferred = cfr.doKeepExact(10000, source, dest, isWater, 144);
+
+		assertEquals("Wrong fluid amount moved", 44, amountTransferred);
+	}
+
+	@Test
+	public void doKeepExact_moves_multiple_valid_fluids() {
+
+		// Create a regulator for testing with, and set it to "Keep Exact" mode
+		CoverFluidRegulator cfr = new CoverFluidRegulator(null, EnumFacing.UP, 0, 1000);
+		cfr.transferMode = TransferMode.KEEP_EXACT;
+
+		IFluidHandler source =
+			new FluidHandlerProxy(
+				new FluidTankList(false),
+				new FluidTankList(false,
+				                  new FluidTank(new FluidStack(FluidRegistry.WATER, 64000), 64000),
+				                  new FluidTank(new FluidStack(FluidRegistry.LAVA, 64000), 64000)));
+
+		// One tank with 100mB water, another with nothing
+		IFluidHandler dest =
+			new FluidHandlerProxy(
+				new FluidTankList(false,
+				                  new FluidTank(new FluidStack(FluidRegistry.WATER, 100), 64000),
+				                  new FluidTank(64000)),
+				new FluidTankList(false));
+
+		// accept any fluid this time
+		int amountTransferred = cfr.doKeepExact(10000, source, dest, fs -> true, 144);
+
+		// expect that 44mB of water and 144mB of lava will be moved
+		assertEquals("Wrong fluid amount moved", 44+144, amountTransferred);
```

Comment:
In this case you should also check content of destination to be really sure that correct amount of both fluids was moved.